flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/SourceWatermarkITCase.scala

```diff
@@ -305,4 +306,56 @@ class SourceWatermarkITCase extends StreamingTestBase {
     assertThat(actualWatermark).isEqualTo(expectedWatermarkOutput)
     assertThat(sink.getAppendResults.sorted).isEqualTo(expectedData.sorted)
   }
+
+  @Test
+  def testWatermarkNotMovingBack(): Unit = {
+    val data = Seq(
+      row(1, LocalDateTime.parse("2024-01-01T00:00:00")),
+      row(3, LocalDateTime.parse("2024-01-03T00:00:00")),
+      row(2, LocalDateTime.parse("2024-01-02T00:00:00"))
+    )
+
+    val dataId = TestValuesTableFactory.registerData(data)
+
+    val ddl =
+      s"""
+         | CREATE Table VirtualTable (
+         |   a INT,
+         |   c TIMESTAMP(3),
+         |   WATERMARK FOR c as c
```

Comment:
This test was intended for `WATERMARK FOR c as SOURCE_WATERMARK()`. Let's implement this test in Java in a fresh class.

Reply:
Any suggestions where to put it? I was looking for a java test I could use, but could not find one. Happy to reimplement all necessary utilities for a fresh test if you feel it's important.

Reply:
I'll try again having it in java.

Reply:
maybe somewhere closer to `org.apache.flink.table.planner.runtime.stream.sql.FunctionITCase` and similar newer generation classes.

Reply:
Migrated it to java.